skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,565 @@
+    """
+    cdef int i, j
+    cdef int center = s / 2
+    # Check if central pixel is too different in the 2 patches
+    cdef float tmp_diff = p1[center, center] - p2[center, center]
+    cdef float init = w[center, center] * tmp_diff * tmp_diff
+    if init > 1:
+        return 0.
+    cdef float distance = 0
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return 0.
+        for j in range(s):
+            tmp_diff = p1[i, j] - p2[i, j]
+            distance += (w[i, j] * tmp_diff * tmp_diff)
+    distance = exp(- distance)
+    return distance
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_2drgb(DTYPE_t [:, :, :] p1,
+                                       DTYPE_t [:, :, :] p2,
+                                       DTYPE_t [:, ::] w, int s):
+    """
+    Compute a Gaussian distance between two image patches.
+
+    Parameters
+    ----------
+    p1 : 3-D array_like
+        first patch, 2D image with last dimension corresponding to channels
+    p2 : 3-D array_like
+        first patch, 2D image with last dimension corresponding to channels
+    w : 2-D array_like
+        array of weigths for the different pixels of the patches
+    s : int
+        linear size of the patches
+
+    Returns
+    -------
+    distance : float
+        Gaussian distance between the two patches
+
+    Notes
+    -----
+    The returned distance is given by
+
+        exp( -w * (p1 - p2)**2)
+    """
+    cdef int i, j
+    cdef int center = s / 2
+    cdef int color
+    cdef float tmp_diff = 0
+    cdef float distance = 0
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return 0.
+        for j in range(s):
+            for color in range(3):
+                tmp_diff = p1[i, j, color] - p2[i, j, color]
+                distance += w[i, j] * tmp_diff * tmp_diff
+    distance = exp(- distance)
+    return distance
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_3d(DTYPE_t [:, :, :] p1,
+                                    DTYPE_t [:, :, :] p2,
+                                    DTYPE_t [:, :, ::] w, int s):
+    """
+    Compute a Gaussian distance between two image patches.
+
+    Parameters
+    ----------
+    p1 : 3-D array_like
+        first patch
+    p2 : 3-D array_like
+        first patch
+    w : 3-D array_like
+        array of weigths for the different pixels of the patches
+    s : int
+        linear size of the patches
+
+    Returns
+    -------
+    distance : float
+        Gaussian distance between the two patches
+
+    Notes
+    -----
+    The returned distance is given by
+
+        exp( -w * (p1 - p2)**2)
+    """
+    cdef int i, j, k
+    cdef float distance = 0
+    cdef float tmp_diff
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
```

Comment:
can you define a constant at the top, e.g. `MAXD`, `MAX_DISTANCE`, or `DISTANCE_CUTOFF`, and use it here? This will make the code a bit more readable.